### PR TITLE
[android] Add hole handling to PolygonOptions

### PR DESCRIPTION
**Pull request by nealsanche**
- Added holes to the Java SDK, JNI, and renderer
- Multiple holes are supported
- Seems stable even if no holes are specified
- Added code to add holes in the Polygon demo app which adds 3 holes to the star over portland.

### CODE REVIEW

**Comment by 1ec5** [comment]
Could you rebase and force-push your changes (fetch upstream master, then rebase the feature branch atop master, then push --force)? Since you last updated your branch, we made a major change down in C++ land (replacing include guards with `#pragma once`, so it's difficult to find your changes in there.

**Comment by nealsanche** [comment]
@1ec5: I did a rebase just now and force pushed.

File: platform/android/src/jni.cpp [hunk 1]
```diff
@@ -303,7 +350,12 @@ std::pair<mbgl::AnnotationSegment, mbgl::ShapeAnnotation::Properties> annotation
     mbgl::AnnotationSegment segment = annotation_segment_from_latlng_jlist(env, points);
     jni::DeleteLocalRef(*env, points);
 
-    return std::make_pair(segment, shapeProperties);
+    jni::jobject* holes = jni::GetField<jni::jobject*>(*env, polygon, *polygonHolesId);
+    mbgl::AnnotationSegments holeSegments = annotation_segments_from_jlist_of_latlng_jlist(env, holes);    
+    jni::DeleteLocalRef(*env, holes);
+    mbgl::Log::Debug(mbgl::Event::JNI, "H-H");
```
Comment (nealsanche) [inline platform/android/src/jni.cpp]: Left in a silly log statement here. Whoops. I'll nuke that guy.

File: platform/android/src/jni.cpp [hunk 2]
```diff
@@ -918,8 +971,8 @@ jni::jarray<jlong>* nativeAddPolygons(JNIEnv *env, jni::jobject* obj, jlong nati
     for (std::size_t i = 0; i < len; i++) {
         jni::jobject* polygon = jni::GetObjectArrayElement(*env, *jarray, i);
 
-        std::pair<mbgl::AnnotationSegment, mbgl::ShapeAnnotation::Properties> segment = annotation_std_pair_from_polygon_jobject(env, polygon);
-        shapes.emplace_back(mbgl::AnnotationSegments { segment.first }, segment.second);
+        std::pair<std::pair<mbgl::AnnotationSegment,mbgl::AnnotationSegments>, mbgl::ShapeAnnotation::Properties> segment = annotation_std_pair_from_polygon_jobject(env, polygon);
```
Comment (nealsanche) [inline platform/android/src/jni.cpp]: I will add a space before the comma.

File: src/mbgl/annotation/shape_annotation_impl.cpp [hunk 3]
```diff
@@ -102,6 +102,23 @@ void ShapeAnnotationImpl::updateTile(const CanonicalTileID& tileID, AnnotationTi
 
         auto ring = geojsonvt::Convert::projectRing(points, tolerance);
         rings.push_back(ring);
+        
+        for (size_t j = 0; j < shape.holeSegments.size(); ++j) {
+            if (shape.holeSegments[j].size() > 0) {
+                std::vector<geojsonvt::LonLat> holePoints;
+                for (size_t i = 0; i < shape.holeSegments[j].size(); ++i) { // try one hole
```
Comment (nealsanche) [inline src/mbgl/annotation/shape_annotation_impl.cpp]: The //try one hole comment was left in by accident. I'll clean that up too.

**Comment by 1ec5** [comment]
Thanks for rebasing, @nealsanche. I’m wondering if you’ve seen #5110, d0ea7c36826554b2433705cd2e62d7420e9d7697 in particular. `mbgl::AnnotationSegments` and `geojsonvt::ProjectedRings` already support holes in polygons: these types consist of an outer ring (the first element), followed by all the inner rings, so no separate “holes” field would be needed at the JNI level on down. I’m hoping to land #5110 today or thereabouts; after that, I think this PR will simplify somewhat.

**Comment by nealsanche** [comment]
@1ec5: You're right, the code would simplify with #5110 in it. I'd be happy to fix this PR when #5110 lands.

**Comment by 1ec5** [comment]
OK, #5110 is merged.

**Comment by nealsanche** [comment]
Alright. I will have a look into porting these changes, likely Monday.

On Sat, May 28, 2016, 12:21 AM Minh Nguyễn <notifications@github.com> wrote:

> OK, #5110 <https://github.com/mapbox/mapbox-gl-native/pull/5110> is
> merged.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/mapbox/mapbox-gl-native/pull/5170#issuecomment-222290973>,
> or mute the thread
> <https://github.com/notifications/unsubscribe/AFHsyafYw6WPosxfXTBe4E8o438621vOks5qF9DEgaJpZM4Iol7e>
> .
>

-- 
Robots and Pencils Inc | 1507 14th Ave SW Calgary AB T3C 0W4
You can unsubscribe from the Robots and Pencils mailing lists by clicking 
here. <https://robotsandpencils.com/contact/unsubscribe.html>

**Comment by nealsanche** [comment]
@1ec5: Alright, I've updated the Android JNI code to use the simplified method of just appending the holes to the shell segment. Seems to be working.

**Comment by nealsanche** [comment]
@1ec5: So, I took a look at the conflicts here today. So much has changed in the JNI layer that I'm not sure where to go with all of this. I'm assuming that this branch is going to dead end, and that the holes API will be implemented in another branch?

**Comment by 1ec5** [comment]
With all the things moving in both mbgl and the Android SDK, unfortunately, that’s probably the case. @bleege can comment on whether polygon hole support is something we can get into an upcoming Android SDK release, or whether it’ll have to wait until the next release cycle. Thank you for taking the time to make this PR, in any case.

**Comment by nealsanche** [comment]
I suspect, even with the current changes, that the code delta to add polygons with holes to Android would  quite small. I would just have to spend a lot more time that one of your team members to make the adaptations.

**Comment by bleege** [comment]
@nealsanche Thanks so much for inquiring about this (and writing code!).  We're currently putting the finishing touches on the Android 4.1.0 release so we're reluctant to add new features this late in that stage.  We'll be starting to think abotu Android 4.2.0 soon though and will be happy to consider it at that time.  Thanks again!

**Comment by 1ec5** [comment]
I made an attempt at reviving this PR in #7397.

**Pull request closed without merging**